File: web/src/main/java/hu/bsstudio/speed/SpeedRaceLapHandler.java
package hu.bsstudio.speed;

import hu.bsstudio.race.speed.SpeedRaceService;
import hu.bsstudio.race.speed.model.SpeedRaceScore;
import hu.bsstudio.team.model.DetailedTeam;
import lombok.RequiredArgsConstructor;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class SpeedRaceLapHandler implements HandlerFunction<ServerResponse> {

    private final SpeedRaceService service;

    @Override
    public Mono<ServerResponse> handle(final ServerRequest request) {
        final var detailedTeam = request.bodyToMono(SpeedRaceScore.class)
            .flatMap(service::updateSpeedRaceOnLap);
        return ServerResponse.ok().body(detailedTeam, DetailedTeam.class);
    }
}

File: server/web/src/main/kotlin/hu/bsstudio/robonaut/speed/SpeedRaceLapHandler.kt
package hu.bsstudio.robonaut.speed

import hu.bsstudio.robonaut.race.speed.SpeedRaceService
import hu.bsstudio.robonaut.race.speed.model.SpeedRaceScore
import org.springframework.web.reactive.function.server.HandlerFunction
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.body
import org.springframework.web.reactive.function.server.bodyToMono
import reactor.core.publisher.Mono

class SpeedRaceLapHandler(private val service: SpeedRaceService) : HandlerFunction<ServerResponse> {
    override fun handle(request: ServerRequest): Mono<ServerResponse> {
        return request.bodyToMono<SpeedRaceScore>()
            .flatMap(service::updateSpeedRaceOnLap)
            .let { detailedTeam -> ServerResponse.ok().body(detailedTeam) }
    }
}
